### ISSUE TITLE
[STORM-1340] Use Travis-CI build matrix to improve test execution times

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5175]
<p>Travis-CI provides a 'build-matrix' that we can use to parallelize the unit tests of various components in storm.<br/>
It also provides a caching mechanism that can reduce the overhead of starting up a test container.</p>

---
<details><summary><i>Originally reported by <a title="Kyle Nusbaum" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=knusbaum">knusbaum</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-1340" target="_blank">Use Travis-CI build matrix to improve test execution times</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Kyle Nusbaum" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=knusbaum">knusbaum</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2016-03-24T20:46:19+00:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user knusbaum opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/902" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/902</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1340" title="Use Travis-CI build matrix to improve test execution times" class="issue-link" data-issue-key="STORM-1340"><del>STORM-1340</del></a>: Use Travis-CI build matrix to improve test execution times</p>

<p>    This patch modifies the Travis-CI config to run storm-core unit tests in parallel with all the external tests, knocking between 4 and 7 minutes off the build time usually.</p>

<p>    It also caches ruby, javascript, and java libraries instead of downloading them each time a container is launched.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/knusbaum/incubator-storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/knusbaum/incubator-storm</a> <a href="https://issues.apache.org/jira/browse/STORM-1340" title="Use Travis-CI build matrix to improve test execution times" class="issue-link" data-issue-key="STORM-1340"><del>STORM-1340</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/902.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/902.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #902</p>

<hr />
<p>commit a0e734727d2335bb0e7a12f8ada8ca07f4908ded<br/>
Author: Kyle Nusbaum <kylejnusbaum@gmail.com><br/>
Date:   2015-11-10T20:27:21Z</p>

<p>    Adding setting to travis.</p>

<p>commit 146a893647267736b01c458dde9a1b0150194b67<br/>
Author: Kyle Nusbaum <kylejnusbaum@gmail.com><br/>
Date:   2015-11-10T22:01:14Z</p>

<p>    Add caching to speed up build.</p>

<p>commit 56297183c86404fb4fe3d623e47808c07b961471<br/>
Author: Kyle Nusbaum <kylejnusbaum@gmail.com><br/>
Date:   2015-11-10T22:17:58Z</p>

<p>    Kicking Travis</p>

<p>commit f0630adb057a4f3dc321e51d14610639753d07e7<br/>
Author: Kyle Nusbaum <kylejnusbaum@gmail.com><br/>
Date:   2015-11-10T22:40:19Z</p>

<p>    Testing travis matrix.</p>

<p>commit 103f71a00a6c41a83349f39a420c2a2d0807ddb7<br/>
Author: Kyle Nusbaum <kylejnusbaum@gmail.com><br/>
Date:   2015-11-10T22:44:24Z</p>

<p>    Testing</p>

<p>    Testing</p>

<p>    Testing</p>

<p>    Testing</p>

<p>    Testing</p>

<p>    Testing</p>

<p>    Testing</p>

<p>    Testing</p>

<p>    Fixing .travis.yml</p>

<p>commit e83678b8fba07f921b661ab0b2a40616726dec55<br/>
Author: Kyle Nusbaum <kylejnusbaum@gmail.com><br/>
Date:   2015-11-11T00:13:43Z</p>

<p>    Cleanup</p>

<p>commit 77417554f1880e113177c874e45ed5721f94b06e<br/>
Author: Kyle Nusbaum <kylejnusbaum@gmail.com><br/>
Date:   2015-11-19T17:31:52Z</p>

<p>    Testing split-up travis.</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/902#discussion_r45664250" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/902#discussion_r45664250</a></p>

<p>    &#8212; Diff: dev-tools/travis/travis-install.sh &#8212;<br/>
    @@ -29,6 +29,7 @@ TRAVIS_SCRIPT_DIR=$( cd "$( dirname "${BASH_SOURCE<span class="error">&#91;0&#93;</span>}" )" && pwd )<br/>
     cd ${STORM_SRC_ROOT_DIR}</p>

<p>     python ${TRAVIS_SCRIPT_DIR}/save-logs.py "install.txt" mvn clean install -DskipTests -Pnative --batch-mode<br/>
    +<br/>
    &#8212; End diff &#8211;</p>

<p>    revert?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/902#issuecomment-159088673" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/902#issuecomment-159088673</a></p>

<p>    Looks good to me drops the build time by about 7 mins. I am +1 if you can address @d2r comment and squash yur changes.  11 commits feels like a bit much.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user knusbaum commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/902#issuecomment-159092312" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/902#issuecomment-159092312</a></p>

<p>    Failure looks unrelated.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/902#issuecomment-159093376" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/902#issuecomment-159093376</a></p>

<p>    Looks good to me +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/902#issuecomment-159346298" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/902#issuecomment-159346298</a></p>

<p>    @knusbaum <br/>
    I'm +1 for the change.<br/>
    Actually I found some builds before rebase still takes 13 min or higher, but mostly it seems to save build time.<br/>
    Great work!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/902" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/902</a></p>
